[FLINK-20491] Support Broadcast State in BATCH execution mode

aljoscha opened the pull request:
Right now, we don't support `DataStream.connect(BroadcastStream)` in `BATCH` execution mode. I believe we can add support for this with not too much work.

The key insight is that we can process the broadcast side before the non-broadcast side. Initially, we were shying away from this because of concerns about `ctx.applyToKeyedState()` which allows the broadcast side of the user function to access/iterate over state from the keyed side. We thought that we couldn't support this. However, since we know that we process the broadcast side first we know that the keyed side will always be empty when doing so. We can thus just make this "keyed iteration" call a no-op, instead of throwing an exception as we do now.

This is work-in-progress. I have yet to add tests.

## Brief change log

* turn broadcast operation into a logical `Transformation`
* add `BATCH` operators for broadcast
* add translators

## Verifying this change

To be done.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? maybe

We need to remove the caveat from the documentation. 🎊

dawidwys commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java:
```diff
@@ -315,23 +342,40 @@ private InputStatus addNextToQueue(HeadElement reuse, DataOutput<IN> output) thr
 		private final CommonContext commonContext;
-		private final int numberOfInputs;
+		private final int numInputs;
+		private final ConcurrentLinkedQueue<Integer> passThroughInputsIndices;
```
Do we need the `Concurrent` queue here? If I am not mistaken all of it happens from within the main mailbox thread.

aljoscha replied:
Could be true, I went for a safe thing but could remove it. 👌

dawidwys commented:
I am good with the general idea and how it looks inside the `MultiInputSortingDataInput.java`. Unless someone objects/has a different opinion, I think we can start properly wiring all the inputs.

dawidwys commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/MultiInputTransformationTranslator.java:
```diff
@@ -54,7 +55,11 @@
 		Collection<Integer> ids = translateInternal(transformation, context);
 		boolean isKeyed = transformation instanceof KeyedMultipleInputTransformation;
```
I think this check is not sufficient now. We should check if any of the inputs has a state selector.

dawidwys commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:
```diff
@@ -45,7 +46,11 @@
 		boolean isKeyed =
 			transformation.getStateKeySelector1() != null && transformation.getStateKeySelector2() != null;
 		if (isKeyed) {
```
Similar to the `MultiInput` shall we account for the situation if only one input is keyed?

aljoscha replied:
I went with the current behaviour, but yes, we could do that. So in that case the non-keyed input would be processed before all the keyed input. Same for the general multi-input.

dawidwys replied:
Yes, I'd do it that way. Otherwise I'd proceed with https://github.com/apache/flink/pull/14324. The current state leads to possibly corrupted results.

aljoscha replied:
I'll do that then. 👌

aljoscha commented:
I updated this with tests and more Javadocs. PTAL.

I still have to add support for multi-input operators and I need to enable mixed keyed/non-keyed operators, as per Dawid's suggestion.

aljoscha commented:
@sjwiesman I'm also adding documentation here, it's a bit terse but still helpful, I think.

sjwiesman commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -237,6 +237,36 @@ next key.
 See [FLIP-140](https://cwiki.apache.org/confluence/x/kDh4CQ) for background
 information on this.
 
+### Order of Processing
+
+The order in which records are processed in operators or user defined functions
+(UDFs) can differ between `BATCH` and `STREAMING` execution.
+
+In `STREAMING` mode, user defined functions should not make any assumptions
+about the order of incoming records. Records are processed immediately once
+they arrive in sources.
+
+In `BATCH` execution mode, there are some operations where Flink guarantees
+order. The ordering can be a side effect of the special task scheduling,
+network shuffle, and state backend (see above) or it can be a conscious choice
+by the system.
+
+There are three general types of input that we can differentiate:
+
+- _keyed input_: input from a `KeyedStream`
+- _broadcast input_: input from a broadcast stream (see also [Broadcast
+  State]({% link dev/stream/state/broadcast_state.md %}))
+- _regular input_: input that isn't any of the above types of input
+
+These are the ordering rules for the different input types
+
+- keyed inputs are processed after all other inputs
+- broadcast inputs are processed before regular inputs
+
+As mentioned above, the keyed input will be grouped and Flink will process all
+records of a keyed group consecutively before processing the next group.
```
@aljoscha What about this? 

```suggestion
### Order of Processing

The order in which records are processed in operators or user-defined functions (UDFs) can differ between `BATCH` and `STREAMING` execution.

In `STREAMING` mode, user-defined functions should not make any assumptions about incoming records' order.
Data is processed as soon as it arrives.

In `BATCH` execution mode, there are some operations where Flink guarantees order. 
The ordering can be a side effect of the particular task scheduling,
network shuffle, and state backend (see above), or a conscious choice by the system.

There are three general types of input that we can differentiate:

- _broadcast input_: input from a broadcast stream (see also [Broadcast
  State]({% link dev/stream/state/broadcast_state.md %}))
- _regular input_: input that isn't any of the above types of input
- _keyed input_: input from a `KeyedStream`

Functions, or Operators, that consume multiple input types will process them in the following order:

- broadcast inputs are processed first
- regular inputs are processed second
- keyed inputs are processed last

For functions that consume from multiple regular or broadcast inputs &mdash; such as a `CoProcessFunction` &mdash; Flink has the right to process data from any input of that type in any order. 
```

aljoscha replied:
Perfect, thanks! I knew I was asking the right person.

kl0u commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java:
```diff
@@ -205,26 +212,46 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
 			final String functionName,
-			final TypeInformation<OUT> outTypeInfo,
-			final TwoInputStreamOperator<IN1, IN2, OUT> operator) {
+			BroadcastProcessFunction<IN1, IN2, OUT> userFunction,
```
Add `final` for uniformity.

kl0u commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java:
```diff
@@ -205,26 +212,46 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
+	@Internal
+	private <OUT> SingleOutputStreamOperator<OUT> transform(
+			final String functionName,
+			KeyedBroadcastProcessFunction<?, IN1, IN2, OUT> userFunction,
```
Add `final` for uniformity.

kl0u commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java:
```diff
@@ -236,26 +263,32 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
+
+	private <KS, OUT> KeyedBroadcastStateTransformation<KS, IN1, IN2, OUT> getBroadcastStateTransformation(
+			final String functionName,
+			KeyedBroadcastProcessFunction<KS, IN1, IN2, OUT> userFunction,
```
Add `final` for uniformity.

kl0u commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java:
```diff
@@ -236,26 +263,32 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
+				outTypeInfo,
+				environment.getParallelism());
+	}
+
```
Now the two `getBroadcastStateTransformation()` methods have no additional logic and each has a single call-site, so why not inlining them in their call-sites?

kl0u commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java:
```diff
@@ -53,111 +50,91 @@
 @Internal
 public class BroadcastStateTransformation<IN1, IN2, OUT> extends PhysicalTransformation<OUT> {
```
The two BroadcastTransformations could have a base class without the key-related fields and the function.

kl0u commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/BroadcastStateTransformation.java:
```diff
@@ -53,111 +50,91 @@
-	}
-
-	public static <IN1, IN2, OUT> BroadcastStateTransformation<IN1, IN2, OUT> forKeyedStream(
+	public static <IN1, IN2, OUT> BroadcastStateTransformation<IN1, IN2, OUT> forStream(
```
We can make the constructor `public` now and remove the static constructors from here and from the `keyed` alternative as now we have two separate classes.

kl0u commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/BroadcastStateTransformationTranslator.java:
```diff
@@ -51,14 +69,18 @@
 		checkNotNull(transformation);
 		checkNotNull(context);
 
+		CoBroadcastWithNonKeyedOperator<IN1, IN2, OUT> operator = new CoBroadcastWithNonKeyedOperator<>(
```
I like that we create the operator here as now we push further down the runtime-related stuff. In the future we could have a translator that gives, for example, not an operator but sth else.

kl0u commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/KeyedBroadcastStateTransformation.java:
```diff
@@ -0,0 +1,160 @@
+
+	private final Transformation<IN2> broadcastStream;
+
+	private final TypeInformation<?> stateKeyType;
```
Now this can have `KS` as a generic type, right? (for type safety)

kl0u commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/KeyedBroadcastStateTransformation.java:
```diff
@@ -0,0 +1,160 @@
+
+	private final TypeInformation<?> stateKeyType;
+
+	private final KeySelector<IN1, ?> keySelector;
```
Same here.

kl0u commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/KeyedBroadcastStateTransformation.java:
```diff
@@ -0,0 +1,160 @@
+	}
+
+	// ------------------------------- Static Constructors -------------------------------
+
```
Now we can remove this and make the constructor `public`, right?

kl0u commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:
```diff
@@ -735,4 +761,10 @@ public int hashCode() {
 			return inputEdge.hashCode();
 		}
 	}
+
```
Why not putting it as a method in the `InputConfig` with a default implementation that returns `false`? I think that this will make the code nicer and also it will avoid checks like `inputConfig instanceof StreamConfig.NetworkInputConfig`

aljoscha replied:
I decided to put it into the `NetworkInputConfig` because that's the only place where it can really be applied right now. But I can also move it to the base class.

kl0u commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamConfig.java:
```diff
@@ -238,10 +231,12 @@ private void setTypeSerializer(String key, TypeSerializer<?> typeWrapper) {
 		}
 	}
 
+	// TODO: this method is a bit misleading, because what it really does is setting up
```
Can't we rename it to sth more descriptive?

kl0u commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java:
```diff
@@ -84,7 +84,7 @@
 	private String transformationUID;
 	private String userHash;
-	private boolean sortedInputs = false;
+	private Map<Integer, StreamConfig.InputRequirement> inputRequirements = new HashMap<>();
```
Why not making this `final`?

Also nit-pick: I find it is nicer to initialize fields in the constructor but feel free to ignore this.

aljoscha replied:
None of the others are initialized in the constructor that's why I went with this one. But yes, I would agree!

kl0u reviewed (changes requested):
Thanks for the work @aljoscha . I had an initial look  and I left some comments. Unfortunately I am not done yet so I will continue with the review.

aljoscha commented:
Thanks for the review! I addressed most comments. I had to force push because the changes would have been a bit hard to follow if I had put them on top of the existing commits.

kl0u commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:
```diff
@@ -143,7 +150,20 @@ else if (configuredInput instanceof StreamConfig.SourceInputConfig) {
+					passThroughIndex++;
+				}
+			}
+			inputs = selectableSortingInputs.getSortedInputs();
```
This line seems to be undoing what you did before, right?

kl0u replied:
Seems like a leftover line but that can lead to bugs.

aljoscha replied:
Yes, it's a leftover. 🙈

dawidwys commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java:
```diff
@@ -90,24 +94,51 @@
-            @SuppressWarnings("unchecked")
+            List<StreamTaskInput<?>> sortedTaskInputs = new ArrayList<>();
+            List<StreamTaskInput<?>> passThroughTaskInputs = new ArrayList<>();
+            int input1Index = 0;
```
It is rather a matter of preference, but I'd move the index computing down below, where it is needed. I found it a bit confusing here and the IDE complains that in line 126 the `passThroughTaskInputs.size()` is always `0`.

We won't need this logic here, if we change the lines 151-160 to:
```StreamTaskInput<?>[] sortedInputs = selectableSortingInputs.getSortedInputs();
StreamTaskInput<?>[] passThroughInputs = selectableSortingInputs.getPassThroughInputs();
if (input1IsSorted) {
    input1 = toTypedInput(sortedInputs[0]);
} else {
    input1 = toTypedInput(passThroughInputs[0]);
}
if (input2IsSorted) {
    input2 = toTypedInput(sortedInputs[sortedInputs.length - 1]);
} else {
    input2 = toTypedInput(passThroughInputs[passThroughInputs.length - 1]);
}
```

We can base that on the assumption that the `input1` will be first in the corresponding group while the `input2` last.

aljoscha replied:
fixing

dawidwys commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamTwoInputProcessorFactory.java:
```diff
@@ -90,24 +94,51 @@
+                passThroughTaskInputs.add(input2);
+            }
+
+            @SuppressWarnings({"unchecked", "rawtypes"})
```
nit: my IDE complains the `"rawtypes"` is unnecessary.

aljoscha replied:
fixed

dawidwys commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/translators/TwoInputTransformationTranslator.java:
```diff
@@ -46,7 +47,11 @@
                 transformation.getStateKeySelector1() != null
                         && transformation.getStateKeySelector2() != null;
         if (isKeyed) {
```
I think we should change it so that we apply sorting if any of the inputs is keyed.

aljoscha replied:
I think we have to yes. I shied away from it here because I would have to add a test that uses the low-level API because right now it's not possible for users to create such transforms.

Could you maybe whip together a test based on the existing test code you already had? We add it here and I change it to correctly forward the requirements?

dawidwys replied:
sure, I will prepare one

dawidwys replied:
I was thinking of tests like those: https://github.com/dawidwys/flink/blob/mix-keyed-2/flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java

aljoscha replied:
Sorry, I also got started on it and should have notified you. 🙈

dawidwys commented on flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java:
```diff
@@ -180,6 +191,107 @@ public void batchSumSingleResultPerKey() throws Exception {
 
+    /** Verifies that all broadcast input is processed before keyed input. */
+    @Test
+    public void batchKeyedBroadcastExecution() throws Exception {
```
nit: How about we test with at least two keys?

aljoscha replied:
adding it to the test 👌

dawidwys reviewed (changes requested):
Thanks a lot for the great work!

A general comment (that I also put inline), I'd be interested to see the impact of the change on benchmarks results.

aljoscha commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -237,6 +237,35 @@ next key.
+
+- _broadcast input_: input from a broadcast stream (see also [Broadcast
+  State]({% link dev/stream/state/broadcast_state.md %}))
+- _regular input_: input that isn't any of the above types of input
```
regular input should be processed before keyed input, because the (sorted) keyed input is always last.

dawidwys replied:
I think I might've not been clear on this one. What I meant is that the line with `regular input` should be below the `keyed input`, because it refers to it. If I understood the sentence correctly it says: `input that isn't broadcast or keyed`, right?

aljoscha replied:
Ah damn, you're right... 🙈 fixing it

aljoscha commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/BroadcastConnectedStream.java:
```diff
@@ -246,28 +246,36 @@ public StreamExecutionEnvironment getExecutionEnvironment() {
+
+        final KeyedBroadcastStateTransformation<KEY, IN1, IN2, OUT> transformation =
+                new KeyedBroadcastStateTransformation<>(
+                        functionName,
```
fixing

aljoscha commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/transformations/AbstractBroadcastStateTransformation.java:
```diff
@@ -0,0 +1,108 @@
+ * Base class for Broadcast State transformations. In a nutshell, this transformation allows to take
+ * a broadcast (non-keyed) stream, connect it with another keyed or non-keyed stream, and apply a
+ * function on the resulting connected stream. This function will have access to all the elements
+ * that belong to the non-keyed, broadcasted side, as this is kept in Flink's state.
```
It's not correct and I copied it from the previous transformation. I can remove the sentence or try and fix it. WDYT?

dawidwys replied:
I'd be ok with simply dropping it.

aljoscha replied:
dropping it

aljoscha commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:
```diff
@@ -142,6 +148,7 @@ public static StreamMultipleInputProcessor create(
                                                             idx, userClassloader))
                                     .toArray(TypeSerializer[]::new),
                             streamConfig.getStateKeySerializer(userClassloader),
+                            new StreamTaskInput[0],
```
It's technically wrong but we don't allow mixed sorted/non-sorted inputs right now, I think.

In general, the support for mixed inputs is in flux with this PR, also relevant for https://github.com/apache/flink/pull/14324.

dawidwys replied:
I don't think there is anything that prohibits mixing keyed with non-keyed inputs. Users can manually craft such Transformation. As for the #14324 didn't we say we will drop it once we merge this PR with a support for mixing such inputs?

BTW, If I remember correctly we had that discussion already ;)

dawidwys replied:
As for the #14324 if I remember correctly we said we will drop it and support mixed keyed/non-keyed inputs: https://github.com/apache/flink/pull/14312#discussion_r540209912

As for the multi inputs those are not exposed in a nice API at all, but I don't think there are any checks or anything that would prohibit users from mixing keyed with non-keyed inputs.

aljoscha replied:
I added support for mixed two-input operations, along with tests. I found some bugs along the way, the fixes are individual commits to make it easier to follow/review.

aljoscha commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java:
```diff
@@ -318,23 +346,40 @@ private InputStatus addNextToQueue(HeadElement reuse, DataOutput<IN> output) thr
         private final CommonContext commonContext;
-        private final int numberOfInputs;
+        private final int numInputs;
+        private final ConcurrentLinkedQueue<Integer> passThroughInputsIndices;
```
changing to a `LinkedList` 👌

aljoscha commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/MultiInputSortingDataInput.java:
```diff
@@ -107,29 +112,38 @@ private MultiInputSortingDataInput(
      */
     public static class SelectableSortingInputs {
```
dropping it.

aljoscha replied:
results are now on http://codespeed.dak8s.net:8000/comparison/

dawidwys commented on docs/dev/datastream_execution_mode.md:
```diff
@@ -1,8 +1,6 @@
-nav-id: datastream_execution_mode
-nav-parent_id: streaming
-nav-pos: 1
+nav-id: datastream_execution_mode nav-parent_id: streaming nav-pos: 1
```
Are all the changes in the file required? Moreover I think this change to the header may break the page.

aljoscha replied:
I don't know how those snuck in... 🙈, reverting

dawidwys commented:
What is the plan for mixed keyed/non-keyed inputs in multiple inputs? If I am not mistaken the implementation in the `StreamMultipleInputProcessorFactory` is not complete and it does not set the pass through inputs correctly.

dawidwys commented on flink-tests/src/test/java/org/apache/flink/api/datastream/DataStreamBatchExecutionITCase.java:
```diff
@@ -180,6 +207,350 @@ public void batchSumSingleResultPerKey() throws Exception {
+
+        try (CloseableIterator<String> resultIterator = result.executeAndCollect()) {
+            List<String> results = CollectionUtil.iteratorToList(resultIterator);
+            // regular, that is non-keyed input is not sorted by timestamp. For keyed inputs
```
Actually the non-broadcasted input is keyed, so we should probably remove the comment.

I copied it over from one of your tests and forgot to remove the comment. The same applies for the other manually crafted two input broadcast operator.

dawidwys reviewed (approved):
+1 % Azure

Looks good now!

aljoscha commented:
Merged this manually
